proto-lens-protoc/src/Data/ProtoLens/Compiler/Definitions.hs

```diff
@@ -355,28 +380,75 @@ messageDefs syntaxType protoPrefix hsPrefix d
             , messageDescriptor = d
             , messageFields =
                   map (liftA2 PlainFieldInfo
-                              (fieldKind syntaxType) (fieldInfo hsPrefix'))
+                              (fieldKind syntaxType mapEntries) (fieldInfo hsPrefix'))
                       $ Map.findWithDefault [] Nothing allFields
             , messageOneofFields = collectOneofFields hsPrefix' d allFields
             , messageUnknownFields =
                   fromString $ "_" ++ hsPrefix' ++ "_unknownFields"
+            , groupFieldNumber = Map.lookup protoName groups
             }
+    subDefs = messageAndEnumDefs
+                    syntaxType
+                    (protoName <> ".")
+                    hsPrefix'
+                    (collectGroupFields $ d ^. field)
+                    (d ^. nestedType)
+                    (d ^. enumType)
+    -- For efficiency, only look for map entries within the immediate
+    -- nested types, rather than recursively searching through all of them.
+    mapEntries = collectMapEntries $ map rootLabel subDefs
+
+-- | If this type is a map entry, retrieves the relevant information
+-- along with the proto name of this type.
+mapEntryInfo :: Definition Name -> Maybe MapEntryInfo
+mapEntryInfo (Message m)
+    | messageDescriptor m ^. options . mapEntry
+    , [keyFd, valueFd] <- messageFields m
+    = Just MapEntryInfo
+                { mapEntryTypeName = messageName m
+                , keyField = plainFieldInfo keyFd
+                , valueField = plainFieldInfo valueFd
+                }
+mapEntryInfo _ = Nothing
+
+-- | Returns a map whose keys are the proto type name of the entry message.
+collectMapEntries :: [(Text, Definition Name)] -> Map.Map Text MapEntryInfo
+collectMapEntries defs =
+    Map.fromList
+        [(protoName, e) | (protoName, d) <- defs, Just e <- [mapEntryInfo d]]
+
+-- | A map whose keys are the proto type names of the groups in a message,
+-- and values are the field numbers for those groups.
+-- (Every group corresponds to exactly one message field.)
+type GroupMap = Map.Map Text Int32
```

Comment:
Idle musing, I wonder if Text for such things is justified or if we would be better off if all this utf8 stuff remained as ByteString of some sort. It's likely impossible to measure any difference between the two.

Reply:
I feel the ability to distinguish between utf8 and raw bytes in the types of the APIs we provide is pretty important.  So I'd want to see a measurable performance issue before strongly considering something else.

There are a handful of Haskell packages that support utf-8-backed Text, e.g `text-short`, but I'm not sure any of them is really in wide use yet.